### PR TITLE
Hotfix/opening townhall(#2516)

### LINKED ISSUE #2516: Crash on opening Minecolonies Townhall
### Minecolonies version

version 0.8.8583 or curse X.X.XX

### Expected behavior

- Townhall GUI opens as usual

### Actual behaviour

- Minecraft crashes

### Steps to reproduce the problem

1. Place the Townhall
2. right click it
3. crash

Crashlog: https://pastebin.com/TX25sF6J

### PR DESCRIPTION
**Pull request by TheYoxy**
Closes #2516

# Changes proposed in this pull request:
- Fixing the bug from #2516
- Moved everything about the happiness from `CitizenManager` to `Colony`

### CODE REVIEW

File: src/main/java/com/minecolonies/coremod/colony/Colony.java [hunk 1]
```diff
@@ -58,142 +58,120 @@
 @SuppressWarnings({Suppression.BIG_CLASS, Suppression.SPLIT_CLASS})
 public class Colony implements IColony
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/Colony.java]: MAJOR Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 38 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
MAJOR class "Colony" has 71 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)

File: src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java [hunk 2]
```diff
@@ -119,8 +104,9 @@ public void sendPackets(
                 if (citizen.getCitizenEntity().isPresent())
                 {
                     final List<EntityCitizen> list = colony.getWorld()
-                            .getEntities(EntityCitizen.class,
-                                    entityCitizen -> entityCitizen.getCitizenColonyHandler().getColony().getID() == colony.getID() && entityCitizen.getCitizenData().getId() == citizen.getId());
+                                                       .getEntities(EntityCitizen.class,
+                                                         entityCitizen -> entityCitizen.getCitizenColonyHandler().getColony().getID() == colony.getID()
+                                                                            && entityCitizen.getCitizenData().getId() == citizen.getId());
 
                     if (!list.isEmpty() && citizen.getCitizenEntity().get().getEntityId() != list.get(0).getEntityId())
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/managers/CitizenManager.java]: MAJOR Call "Optional#isPresent()" before accessing the value. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3655)

File: src/main/java/com/minecolonies/coremod/colony/ColonyView.java [hunk 3]
```diff
@@ -39,79 +39,71 @@
 public final class ColonyView implements IColony
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyView.java]: MAJOR class "ColonyView" has 60 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)

File: src/main/java/com/minecolonies/coremod/colony/ColonyView.java [hunk 4]
```diff
@@ -39,79 +39,71 @@
...
+    /**
+     * Check if the colony has a warehouse.
+     */
+    private       boolean                             hasColonyWarehouse;
     /**
      * The overall happiness of the colony.
      */
-    private double overallHappiness = 5;
-
+    private       double                              overallHappiness   = 5;
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyView.java]: MAJOR Assign this magic number 5 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 5]
```diff
@@ -1149,6 +1129,24 @@ private static File getBackupSaveLocation(final Date date)
...
+    public static void addToZipFile(final String fileName, final ZipOutputStream zos, final File folder)
+    {
+        final File file = new File(folder, fileName);
+        try (FileInputStream fis = new FileInputStream(file))
+        {
+            zos.putNextEntry(new ZipEntry(fileName));
+            Files.copy(file, zos);
+            fis.close();
+        }
+        catch (final Exception e)
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java]: MAJOR Either log or rethrow this exception. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1166)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 6]
```diff
@@ -61,56 +61,46 @@
  */
 public final class ColonyManager
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java]: MAJOR Split this class into smaller and more specialized ones to reduce its dependencies on other classes from 39 to the maximum authorized 30 or less. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1200)
MAJOR class "ColonyManager" has 59 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)

File: src/main/java/com/minecolonies/coremod/colony/ColonyManager.java [hunk 7]
```diff
@@ -1053,60 +1073,20 @@ public static boolean backupColonyData()
...
      * @param compound NBT Tag.
      */
     public static void readFromNBT(@NotNull final NBTTagCompound compound, @NotNull final World world)
     {
-        if(!compound.hasKey(TAG_DISTANCE))
+        if (!compound.hasKey(TAG_DISTANCE))
         {
             Configurations.gameplay.workingRangeTownHallChunks =
-                    (int) ((Math.cos(45.0 / HALF_A_CIRCLE * Math.PI) * Configurations.gameplay.workingRangeTownHall) / BLOCKS_PER_CHUNK);
+              (int) ((Math.cos(45.0 / HALF_A_CIRCLE * Math.PI) * Configurations.gameplay.workingRangeTownHall) / BLOCKS_PER_CHUNK);
```
Comment (BuildTeamLinter) [inline src/main/java/com/minecolonies/coremod/colony/ColonyManager.java]: MAJOR Assign this magic number 45.0 to a well-named constant, and use the constant instead. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS109)

**Comment by BuildTeamLinter** [comment]
SonarQube analysis reported 76 issues
* ![CRITICAL][CRITICAL] 2 critical
* ![MAJOR][MAJOR] 67 major
* ![MINOR][MINOR] 5 minor
* ![INFO][INFO] 2 info

Watch the comments in this conversation to review them.

#### Top 10 extra issues

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. ![CRITICAL][CRITICAL] [WindowTownHall.java#L1027](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L1027): Add a default case to this switch. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3ASwitchLastCaseIsDefaultCheck)
1. ![CRITICAL][CRITICAL] [ColonyManager.java#L1004](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L1004): Refactor this code to not nest more than 5 if/for/while/switch/try statements. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS134)
1. ![MAJOR][MAJOR] [blocks#](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/blocks#): Add a 'package-info.java' file to document the 'blocks' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [gui#](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/client/gui#): Add a 'package-info.java' file to document the 'gui' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L49](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L49): class "WindowTownHall" has 41 methods, which is greater than the 35 authorized. Split it into smaller classes. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1448)
1. ![MAJOR][MAJOR] [WindowTownHall.java#L894](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/client/gui/WindowTownHall.java#L894): Reduce this anonymous class number of lines from 41 to at most 40, or make it a named class. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1188)
1. ![MAJOR][MAJOR] [render#](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/client/render#): Add a 'package-info.java' file to document the 'render' package [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS1228)
1. ![MAJOR][MAJOR] [Colony.java#L527](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/colony/Colony.java#L527): The Cyclomatic Complexity of this method "onWorldTick" is 16 which is greater than 15 authorized. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
1. ![MAJOR][MAJOR] [ColonyManager.java#L385](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L385): Provide multiple methods instead of using "canDestroy" to determine which action to take. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS2301)
1. ![MAJOR][MAJOR] [ColonyManager.java#L661](https://github.com/Minecolonies/minecolonies/blob/577457f12b7e78b6d899b794384dec25f5f8a5ef/src/main/java/com/minecolonies/coremod/colony/ColonyManager.java#L661): Refactor this method to reduce its Cognitive Complexity from 21 to the 20 allowed. [rule](https://sonar.minecolonies.com/coding_rules#rule_key=squid%3AS3776)

[CRITICAL]: https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL'
[INFO]: https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO'
[MAJOR]: https://sonarsource.github.io/sonar-github/severity-major.png 'Severity: MAJOR'
[MINOR]: https://sonarsource.github.io/sonar-github/severity-minor.png 'Severity: MINOR'

**Review by marchermans — DISMISSED** [review]

**Pull request closed without merging**